src/main/restartDialog.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * Dialog for the "Restart Needed" functionality
+ *
+ * Introduces an Electron BrowserWindow component that manages the "Restart Needed" pop-up, which includes the creation and connection of the dialog.
+ * Copyright 2016 Steven Githens
+ * Copyright 2016-2017 OCAD University
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+"use strict";
+
+var fluid   = require("infusion");
+var gpii    = fluid.registerNamespace("gpii");
+var ipcMain = require("electron").ipcMain;
+
+require("./dialog.js");
+
+/**
+ * A component that serves as simple interface for communication with the
+ * electron `BrowserWindow` restart dialog.
+ */
+fluid.defaults("gpii.app.dialog.restartDialog.channel", {
+    gradeNames: ["fluid.component"],
+
+    events: {
+        onClosed: null // provided by parent component
+    },
+
+    listeners: {
+        "onCreate.registerChannel": {
+            funcName: "gpii.app.dialog.restartDialog.channel.register",
+            args: ["{that}.events"]
+        }
+    },
+
+    invokers: {
+        updatePendingChanges: {
+            funcName: "gpii.app.notifyWindow",
+            args: [
+                "{dialog}.dialog",
+                "onRestartRequired",
+                "{arguments}.0"
+            ]
+        }
+    }
+});
+
+/**
+ * Register for events from the managed Electron `BrowserWindow` (the renderer process).
+ * @param events {Objects} Events that are to be mapped to dialog actoins
+ */
+gpii.app.dialog.restartDialog.channel.register = function (events) {
+    /*
+     * NOTE: all other dialog events - onRestartNow, onRestartLater, onUndoChnages
+     * are handled by the `gpii.app.psp` IPC handler as they share both share the same
+     * functionality and ipcMain channel.
+     */
+
+    ipcMain.on("onClosed", function (/*event, message*/) {
+        events.onClosed.fire();
+    });
+};
+
+
+/**
+ * Handles logic for the dialog popup for the "Restart required" functionality.
+ * Creates an Electron `BrowserWindow` and manages it.
+ */
+fluid.defaults("gpii.app.dialog.restartDialog", {
+    gradeNames: ["gpii.app.dialog"],
+
+    invokers: {
+        showIfNeeded: {
+            funcName: "gpii.app.dialog.restartDialog.showIfNeeded",
+            args: ["{that}", "{arguments}.0"]
+        },
+        isShown: {
```

Comment:
Rather than reaching into the impl here, make this a standard modelised element of the base class. That is, place "isShown" into a model field, i) isShown invoker is removed in favour of the client examining this field, ii) show and hide are just model listeners rather than invokers

Reply:
Our assumption for having a show/hide invokers was that the logic for showing/hiding of a dialog could be more complex. Take for instance the PSP itself (even though it is not a dialog). When showing it, we actually move it to the screen, focus it and change a property in the model. And what if there is a need to pass a parameter when showing a dialog (like when displaying a survey)? Should we have an invoker in that case that sets the parameter to the component’s options or model, then changes the `isShown` model property and leaves it up to the model listener to handle the showing operation taking into account the set parameter?

Maybe we are accustomed to using functions/methods for such operations in other frameworks/programming languages instead of modifying the intrinsic object properties. Is this not applicable in this particular case for Infusion components?

Reply:
@amb26, I just wanted to check if you have had the chance to review Georgi's comment. It's the largest remaining comment from the original PR and I wanted to get some sense on the remaining work depending on what needs to be done here. Thank you!